symengine/polynomial.h

```diff
@@ -235,156 +132,307 @@ class UnivariateExprPolynomial
-        : poly_(UnivariatePolynomial::create(symbol(varname), {0, 1}))
+    UnivariateExprPolynomial(const std::string &s) : dict_({{1, Expression(1)}})
     {
     }
-    UnivariateExprPolynomial(RCP<const UnivariatePolynomial> p)
-        : poly_(std::move(p))
+    UnivariateExprPolynomial(const map_int_Expr &p)
     {
+        dict_ = p;
+        auto iter = dict_.begin();
+        while (iter != dict_.end()) {
+            if (Expression(0) == iter->second) {
+                auto toErase = iter;
+                iter++;
+                dict_.erase(toErase);
+            } else
+                iter++;
+        }
     }
-    UnivariateExprPolynomial(Expression expr)
-        : poly_(UnivariatePolynomial::create(symbol(""), {expr}))
+    UnivariateExprPolynomial(const Expression &expr)
     {
+        if (expr != Expression(0))
+            dict_ = {{0, std::move(expr)}};
     }
+
     UnivariateExprPolynomial &operator=(const UnivariateExprPolynomial &)
         = default;
     UnivariateExprPolynomial &
     operator=(UnivariateExprPolynomial &&other) SYMENGINE_NOEXCEPT
     {
         if (this != &other)
-            this->poly_ = std::move(other.poly_);
+            this->dict_ = std::move(other.dict_);
         return *this;
     }
 
     friend std::ostream &operator<<(std::ostream &os,
                                     const UnivariateExprPolynomial &expr)
     {
-        os << expr.poly_->__str__();
+        os << expr.dict_;
         return os;
     }
 
     friend UnivariateExprPolynomial operator+(const UnivariateExprPolynomial &a,
                                               const UnivariateExprPolynomial &b)
     {
-        return UnivariateExprPolynomial(add_uni_poly(*a.poly_, *b.poly_));
+        UnivariateExprPolynomial c = a;
+        c += b;
+        return UnivariateExprPolynomial(std::move(c));
     }
 
     UnivariateExprPolynomial &operator+=(const UnivariateExprPolynomial &other)
     {
-        poly_ = add_uni_poly(*poly_, *other.poly_);
+        for (auto &it : other.dict_)
+            dict_[it.first] += it.second;
         return *this;
     }
 
     friend UnivariateExprPolynomial operator-(const UnivariateExprPolynomial &a,
                                               const UnivariateExprPolynomial &b)
     {
-        return UnivariateExprPolynomial(sub_uni_poly(*a.poly_, *b.poly_));
+        UnivariateExprPolynomial c = a;
+        c -= b;
+        return UnivariateExprPolynomial(std::move(c));
     }
 
     UnivariateExprPolynomial operator-() const
     {
-        return neg_uni_poly(*this->poly_);
+        map_int_Expr dict;
+        for (auto &it : dict_)
+            dict[it.first] = -it.second;
+        return UnivariateExprPolynomial(dict);
     }
 
     UnivariateExprPolynomial &operator-=(const UnivariateExprPolynomial &other)
     {
-        poly_ = sub_uni_poly(*poly_, *other.poly_);
+        for (auto &it : other.dict_)
+            dict_[it.first] -= it.second;
```

Comment:
Same point on canonicalization here. I think just putting a check `dict_[it.first] - it.second != 0` should suffice?